pyiceberg/avro/resolver.py

```diff
@@ -233,7 +256,93 @@ def skip(self, decoder: BinaryDecoder) -> None:
         pass
 
 
-class SchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Reader]):
+class WriteSchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Writer]):
+    def schema(self, write_schema: Schema, data_schema: Optional[IcebergType], result: Writer) -> Writer:
+        return result
+
+    def struct(self, write_schema: StructType, data_struct: Optional[IcebergType], field_writers: List[Writer]) -> Writer:
+        if not isinstance(data_struct, StructType):
+            raise ResolveError(f"File/write schema are not aligned for struct, got {data_struct}")
+
+        data_positions: Dict[int, int] = {field.field_id: pos for pos, field in enumerate(data_struct.fields)}
+        results: List[Tuple[Optional[int], Writer]] = []
+
+        for writer, write_field in zip(field_writers, write_schema.fields):
+            if write_field.field_id in data_positions:
+                results.append((data_positions[write_field.field_id], writer))
+            else:
+                # There is a default value
+                if write_field.write_default is not None:
+                    # The field is not in the record, but there is a write default value
+                    results.append((None, DefaultWriter(writer=writer, value=write_field.write_default)))  # type: ignore
+                elif write_field.required:
+                    raise ValueError(f"Field is required, and there is no write default: {write_field}")
+
+        return StructWriter(field_writers=tuple(results))
+
+    def field(self, write_field: NestedField, data_type: Optional[IcebergType], field_writer: Writer) -> Writer:
+        return field_writer if write_field.required else OptionWriter(field_writer)
+
+    def list(self, write_list_type: ListType, write_list: Optional[IcebergType], element_reader: Writer) -> Writer:
```

Comment:
Nit: `element_reader` should be `element_writer`. Same for the `map` method args.

Reply:
Nice, thanks!